File: day1/part2.js
import { readFileSync } from "node:fs";

const numDict = {
  o: ["one"],
  t: ["two", "three"],
  f: ["four", "five"],
  s: ["six", "seven"],
  e: ["eight"],
  n: ["nine"],
};

const numTable = {
  one: 1,
  two: 2,
  three: 3,
  four: 4,
  five: 5,
  six: 6,
  seven: 7,
  eight: 8,
  nine: 9,
};

function isDigit(c) {
  return /^\d$/.test(c);
}

function interpret(s) {
  const numbers = [];
  for (let i = 0; i < s.length; i++) {
    if (isDigit(s[i])) {
      numbers.push(Number.parseInt(s[i], 10));
      continue;
    }
    if (s[i] in numDict) {
      numDict[s[i]]
        .filter((num) => s.substring(i, i + num.length) === num)
        .forEach((num) => numbers.push(numTable[num]));
    }
  }
  return numbers;
}

function parseData(data) {
  data = data.split("\n");
  return data;
}

function solve(data){
  data = parseData(data);
  return data.reduce((total, current) => {
    const legalNumbers = interpret(current).map((num) =>
      Number.parseInt(num, 10)
    );
    return total + legalNumbers[0] * 10 + legalNumbers.at(-1);
  }, 0);
}

const data = readFileSync("./day1/input.txt", "utf-8");
console.log(solve(data));

File: day1/part2.py
import functools
import os

input_path = "input.txt"
with open(
    os.path.join(os.path.dirname(os.path.realpath(__file__)), input_path), "r"
) as f:
    lines = f.read().splitlines()

num_dict = {
    "o": ["one"],
    "t": ["two", "three"],
    "f": ["four", "five"],
    "s": ["six", "seven"],
    "e": ["eight"],
    "n": ["nine"],
}

num_table = {
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
}


def line_to_num(line: str) -> list[int]:
    nums = []
    for idx, char in enumerate(line):
        if char.isdigit():
            nums.append(int(char))
        if char in num_dict:
            nums.extend(
                [
                    num_table[num]
                    for num in num_dict[char]
                    if line[idx : idx + len(num)] == num
                ]
            )
    return nums


line_nums = list(map(line_to_num, lines))


def num_sum(acc: int, nums: list[int]) -> int:
    return acc + nums[0] * 10 + nums[-1]


print(functools.reduce(num_sum, line_nums, 0))
